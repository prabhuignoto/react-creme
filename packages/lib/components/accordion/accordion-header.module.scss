@use '../../design/theme.scss';
@use '../../design/core.scss';

$sizes: (sm, md, lg);

.header {
  @extend %left;
  align-items: center;
  color: theme.$primary;
  height: var(--rc-accordion-header-height, 35px);
  justify-items: flex-start;
  outline: 0;
  padding-left: 0.5rem;
  position: relative;
  text-transform: capitalize;

  &:not(.header-icon) {
    display: grid;
    grid-template-columns: 2rem calc(100% - 2.5rem) 0.5rem;
  }

  &.colorize {
    background: rgba(theme.$secondary-rgb, 0.5);
  }

  &.full_width {
    width: calc(100% - 0.5rem);
  }
}

.header:not(.disable-collapse):hover {
  @extend %border-radius;
  cursor: pointer;

  &.dark {
    background: theme.$gunmetal-gray;
  }

  &:not(.dark) {
    background: rgba(theme.$secondary-rgb, 0.25);
  }
}

.selected {
  @extend %border-radius;

  &.dark {
    background: theme.$gunmetal-gray;
  }

  &:not(.dark) {
    background: rgba(theme.$secondary-rgb, 0.25);
  }
}

.title {
  @extend %left;
  display: block;
  white-space: nowrap;

  @each $size in $sizes {
    &.title_#{$size} {
      @extend %text-#{$size};
      line-height: #{$size}-line-height;
    }
  }

  &:not(.title_bold) {
    font-weight: 400;
  }

  &.title_bold {
    font-size: 1.1rem;
    font-weight: 500;
  }

  &.colorize {
    color: theme.$primary;
  }
}

.icon {
  @extend %icon;
  cursor: pointer;
  transition: transform 0.2s ease-in;

  &:not(.colorize) {
    color: var(--icon-color, theme.$primary);
  }

  &.colorize {
    color: theme.$primary;
  }

  @each $size in $sizes {
    &.icon-#{$size} {
      @extend %icon-#{$size};
    }
  }

  &.icon-open:not(.icon-plus):not(.custom-icon) {
    transform: rotate(90deg);
  }

  &.disable-icon {
    visibility: hidden;
  }
}
