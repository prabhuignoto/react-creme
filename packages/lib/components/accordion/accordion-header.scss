@use '../../design/theme.scss';
@use '../../design/layout.scss';

$header-height: 35px;
$sizes: (sm, md, lg);

.rc-accordion-header {
  @extend %left;
  align-items: center;
  color: theme.$primary;
  height: $header-height;
  justify-items: center;
  outline: 0;
  position: relative;
  text-transform: capitalize;

  &:not(.rc-accordion-disable-collapse):hover {
    @extend %border-radius;
    background: rgba(theme.$secondary-rgb, 0.25);
    cursor: pointer;
  }

  &:not(.rc-accordion-header-icon) {
    display: grid;
    grid-template-columns: 2rem calc(100% - 2.5rem) 0.5rem;
  }

  &.rc-accordion-selected {
    @extend %border-radius;
    background: rgba(theme.$secondary-rgb, 0.25);
  }
}

.rc-accordion-title {
  @extend %left;
  display: block;
  white-space: nowrap;

  @each $size in $sizes {
    &.rc-accordion-title-#{$size} {
      @extend %text-#{$size};
      line-height: #{$size}-line-height;
    }
  }

  &:not(.rc-accordion-title-bold) {
    font-weight: 400;
  }

  &.rc-accordion-title-bold {
    font-weight: 500;
  }
}

.rc-accordion-icon {
  @extend %icon;
  color: var(--icon-color, theme.$primary);
  cursor: pointer;
  transition: transform 0.2s ease-in;

  @each $size in $sizes {
    &.rc-accordion-icon-#{$size} {
      @extend %icon-#{$size};
    }
  }

  &.rc-accordion-icon-open:not(.rc-accordion-icon-plus):not(.rc-accordion-custom-icon) {
    transform: rotate(90deg);
  }

  &.rc-accordion-disable-icon {
    visibility: hidden;
  }
}
