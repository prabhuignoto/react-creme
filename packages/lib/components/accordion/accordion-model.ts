import { ReactNode } from 'react';

export type AccordionProps = {
  alignIconRight?: boolean;
  autoSetBodyHeight?: boolean;
  border?: boolean;
  children?: ReactNode;
  customContent?: React.ReactNode;
  customIcon?: ReactNode;
  disableCollapse?: boolean;
  disableIcon?: boolean;
  expanded?: boolean | null;
  focusable?: boolean;
  iconColor?: string;
  iconType?: 'chevron' | 'plus';
  id?: string;
  isTitleBold?: boolean;
  maxHeight?: number;
  onChange?: (open: boolean) => void;
  onCollapsed?: (id: string) => void;
  onExpanded?: (id: string) => void;
  onRendered?: () => void;
  selected?: boolean;
  title?: string;
  titleColor?: string;
  transition?: string;
};

export type AccordionGroupProps = Pick<
  AccordionProps,
  | 'disableCollapse'
  | 'alignIconRight'
  | 'iconColor'
  | 'titleColor'
  | 'isTitleBold'
  | 'focusable'
> & {
  autoClose?: boolean;
  border?: boolean;
  children?: ReactNode[];
  disableIcon?: boolean;
  expanded?: boolean;
  iconType?: 'chevron' | 'plus';
  icons?: React.ReactNode[];
  titles?: string[];
};

export interface AccordionItemProps {
  expanded?: boolean;
  focusable?: boolean;
  id?: string;
}

export type AccordionHeaderProps = Pick<
  AccordionProps,
  | 'disableIcon'
  | 'focusable'
  | 'alignIconRight'
  | 'disableCollapse'
  | 'iconType'
  | 'iconColor'
  | 'title'
  | 'customIcon'
  | 'isTitleBold'
  | 'selected'
  | 'customContent'
> & {
  accordionBodyId?: string;
  accordionId?: string;
  onToggle?: () => void;
  open?: boolean | null;
};
