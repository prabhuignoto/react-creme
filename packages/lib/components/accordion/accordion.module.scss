@use '@design/core.scss';
@use '@design/theme.scss';

$transition: max-height 0.5s var(--transition), opacity 0.5s var(--transition);

.accordion {
  @extend %col;
  align-items: flex-start;
  width: 100%;

  &:not(.no-border) {
    @extend %border;
  }
}

.align-icon-rt {
  direction: rtl;
}

.body {
  margin: 0 auto;
  overflow: hidden;
  width: 100%;

  &.animate {
    transition: $transition;
  }

  &.open {
    max-height: var(--max-height, 100%);
    opacity: 1;
  }

  &.close {
    max-height: 0;
    opacity: 0;
  }
}
