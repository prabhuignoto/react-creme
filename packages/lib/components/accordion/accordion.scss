@use '../../design/layout.scss';
@use '../../design/colors.scss';
@use '../../design/icon.scss';
// @use "../../design/animate.scss";

$header-height: 35px;
$transition: max-height 0.5s var(--transition), opacity 0.5s var(--transition);

.rc-accordion {
  @extend %col;
  align-items: flex-start;
  width: 100%;

  &:not(.rc-accordion-no-border) {
    @extend %border;
  }
}

.rc-accordion-header {
  @extend %left;
  align-items: center;
  color: var(--title-color);
  height: $header-height;
  justify-items: center;
  justify-items: center;
  outline: 0;
  position: relative;
  text-transform: capitalize;

  &:not(.rc-accordion-disable-collapse):hover {
    @extend %border-radius;
    background: rgba(colors.$secondary-rgb, 0.25);
    cursor: pointer;
  }

  &:not(.rc-accordion-disable-icon):not(.rc-accordion-header-icon) {
    display: grid;
    grid-template-columns: 2rem calc(100% - 2.5rem) 0.5rem;
  }
}

.rc-accordion-align-icon-rt {
  direction: rtl;
}

.rc-accordion-body {
  margin: 0 auto;
  max-height: var(--max-height);
  overflow: hidden;
  width: 100%;

  &.rc-accordion-open {
    opacity: 1;
    transition: $transition;
  }

  &.rc-accordion-close {
    max-height: 0;
    opacity: 0;
    transition: $transition;
  }
}

.rc-accordion-title {
  @extend %left;
  display: block;
  font-size: 1rem;
  padding-left: 0.5rem;
  white-space: nowrap;

  &:not(.rc-accordion-title-bold) {
    font-weight: 400;
  }

  &.rc-accordion-title-bold {
    font-weight: 500;
  }
}

.rc-accordion-icon {
  @extend %icon-md;
  color: var(--icon-color, colors.$primary);
  cursor: pointer;
  transition: transform 0.2s ease-in;

  &.rc-accordion-icon-open:not(.rc-accordion-icon-plus):not(.rc-accordion-custom-icon) {
    transform: rotate(90deg);
  }
}
