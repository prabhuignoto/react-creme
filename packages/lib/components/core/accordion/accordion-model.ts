import { ReactNode } from 'react';

export type AccordionProps = {
  alignIconRight?: boolean;
  animate?: boolean;
  autoSetBodyHeight?: boolean;
  border?: boolean;
  children?: ReactNode;
  colorizeHeader?: boolean;
  customContent?: React.ReactNode;
  customIcon?: ReactNode;
  disableARIA?: boolean;
  disableCollapse?: boolean;
  disableIcon?: boolean;
  expanded?: boolean | null;
  focusable?: boolean;
  fullWidth?: boolean;
  headerHeight?: number;
  iconColor?: string;
  iconType?: 'chevron' | 'plus';
  id?: string;
  isTitleBold?: boolean;
  maxHeight?: number;
  onChange?: (open: boolean) => void;
  onCollapsed?: (id: string) => void;
  onExpanded?: (id: string) => void;
  onRendered?: () => void;
  selected?: boolean;
  size?: 'sm' | 'md' | 'lg';
  title?: string;
  titleColor?: string;
  transition?: string;
};

export type AccordionGroupProps = Pick<
  AccordionProps,
  | 'disableCollapse'
  | 'alignIconRight'
  | 'iconColor'
  | 'titleColor'
  | 'isTitleBold'
  | 'focusable'
  | 'disableARIA'
  | 'size'
  | 'fullWidth'
  | 'colorizeHeader'
  | 'headerHeight'
> & {
  autoClose?: boolean;
  border?: boolean;
  children?: ReactNode[];
  disableIcon?: boolean;
  expanded?: boolean;
  iconType?: 'chevron' | 'plus';
  icons?: React.ReactNode[];
  titles?: string[];
};

export interface AccordionItemProps {
  expanded?: boolean;
  focusable?: boolean;
  id?: string;
}

export type AccordionHeaderProps = Pick<
  AccordionProps,
  | 'disableIcon'
  | 'focusable'
  | 'alignIconRight'
  | 'disableCollapse'
  | 'iconType'
  | 'iconColor'
  | 'title'
  | 'customIcon'
  | 'isTitleBold'
  | 'selected'
  | 'customContent'
  | 'size'
  | 'fullWidth'
  | 'headerHeight'
> &
  Pick<AccordionGroupProps, 'colorizeHeader'> & {
    accordionBodyId?: string;
    accordionId?: string;
    onToggle?: () => void;
    open?: boolean | null;
  };
