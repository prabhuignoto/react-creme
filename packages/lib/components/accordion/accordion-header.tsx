import classnames from "classnames";
import React, { useMemo, useRef } from "react";
import { ChevronRightIcon, MinusIcon, PlusIcon } from "../../icons";
import { useFocus } from "../common/effects/useFocus";
import { AccordionHeaderProps } from "./accordion-model";

const AccordionHeader: React.FunctionComponent<AccordionHeaderProps> = ({
  disableIcon,
  focusable,
  alignIconRight,
  disableCollapse,
  accordionBodyId,
  isTitleBold,
  title,
  customIcon,
  iconType,
  onToggle,
  accordionId,
  open,
}) => {
  const accordionHeaderClass = useMemo(
    () =>
      classnames("rc-accordion-header", {
        "rc-accordion-align-icon-rt": alignIconRight,
        "rc-accordion-disable-icon": disableIcon,
        "rc-accordion-focusable": focusable,
      }),
    [alignIconRight, focusable]
  );

  const ref = useRef(null);

  const focusProps = useMemo(
    () => (focusable && !disableCollapse ? { tabIndex: 0 } : null),
    [focusable, disableCollapse]
  );

  const collapsibleProps = useMemo(() => {
    return !disableCollapse
      ? {
          "aria-controls": accordionBodyId,
          "aria-expanded": !!open,
          onClick: onToggle,
        }
      : null;
  }, []);

  const titleClass = useMemo(() => {
    return classnames("rc-accordion-title", {
      "rc-accordion-title-bold": isTitleBold,
    });
  }, [isTitleBold]);

  const iconClass = useMemo(() => {
    const classes: string[] = [];

    return classnames([...classes, "rc-accordion-icon"], {
      "rc-accordion-custom-icon": customIcon,
      [`rc-accordion-icon-${iconType}`]: true,
      "rc-accordion-icon-open": open,
    });
  }, [open, customIcon]);

  const icon = useMemo(() => {
    if (disableIcon) {
      return null;
    }

    if (customIcon) {
      return customIcon;
    } else if (iconType === "chevron") {
      return <ChevronRightIcon />;
    } else if (iconType === "plus") {
      return open ? <MinusIcon /> : <PlusIcon />;
    }
  }, [iconType, open, disableIcon]);

  if (focusable) {
    useFocus(ref, onToggle);
  }

  return (
    <div
      className={accordionHeaderClass}
      id={accordionId}
      ref={ref}
      role="button"
      {...focusProps}
      {...collapsibleProps}
    >
      {!disableIcon && <span className={iconClass}>{icon}</span>}
      <span className={titleClass}>{title}</span>
    </div>
  );
};

export { AccordionHeader };
