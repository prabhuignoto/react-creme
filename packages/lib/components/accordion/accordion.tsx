import cls from 'classnames';
import { nanoid } from 'nanoid';
import * as React from 'react';
import {
  CSSProperties,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { useFirstRender } from '../common/effects/useFirstRender';
import { AccordionHeader } from './accordion-header';
import { AccordionProps } from './accordion-model';
import './accordion.scss';

const Accordion: React.FunctionComponent<AccordionProps> = React.memo(
  ({
    alignIconRight = false,
    autoSetBodyHeight = true,
    border = false,
    children,
    customContent,
    customIcon = null,
    disableCollapse = false,
    disableIcon = false,
    expanded = null,
    focusable = true,
    iconColor,
    iconType = 'chevron',
    id,
    isTitleBold = false,
    onChange,
    onCollapsed,
    onExpanded,
    onRendered,
    selected = false,
    title,
    titleColor = '#000',
    transition = 'cubic-bezier(0.19, 1, 0.22, 1)',
    disableARIA,
    size = 'sm',
    animate = true,
  }: AccordionProps) => {
    const accordionID = useRef(id || `accordion-${nanoid()}`);
    const accordionBodyId = useRef(`accordion-body-${nanoid()}`);

    const ref = useRef<HTMLDivElement | null>(null);

    const [open, setOpen] = useState(expanded);
    const [bodyHeight, setBodyHeight] = useState(0);

    const toggleAccordion = useCallback(() => {
      enableCallback.current = true;

      setOpen(prev => {
        onChange?.(!prev);
        return !prev;
      });
    }, [open]);

    const enableCallback = useRef(false);
    const isFirstRender = useFirstRender();

    const accordionBodyClass = useMemo(
      () =>
        cls('rc-accordion-body', {
          'rc-accordion-animate': animate,
          'rc-accordion-close': !open,
          'rc-accordion-open': open && !isFirstRender.current,
        }),
      [open]
    );

    const style = useMemo(
      () => ({
        '--icon-color': iconColor,
        '--title-color': titleColor,
        '--transition': transition,
      }),
      [open, bodyHeight]
    );

    const styleWithHeight = useMemo(
      () =>
        style && !disableCollapse
          ? {
              ...style,
              '--max-height': open
                ? bodyHeight
                  ? `${bodyHeight}px`
                  : `${100}px`
                : '0px',
            }
          : {},
      [autoSetBodyHeight, bodyHeight, style, open]
    );

    const accordionClass = useMemo(
      () =>
        cls('rc-accordion', {
          'rc-accordion-no-border': !border,
          'rc-accordion-open': open,
        }),
      [border, open, alignIconRight]
    );

    const onInitRef = useCallback(node => {
      if (node) {
        ref.current = node as HTMLDivElement;
      }
    }, []);

    useEffect(() => {
      if (isFirstRender.current || !enableCallback.current) {
        return;
      }

      if (open) {
        onExpanded && onExpanded(accordionID.current);
      } else {
        onCollapsed && onCollapsed(accordionID.current);
      }
    }, [open]);

    useEffect(() => {
      enableCallback.current = false;

      if (expanded !== null) {
        setOpen(expanded);
      }
    }, [expanded]);

    useEffect(() => {
      if (children) {
        const height = ref.current?.scrollHeight;

        if (height && autoSetBodyHeight) {
          setBodyHeight(height);
        }

        onRendered?.();
      }
    }, [children]);

    const ARIA = useMemo(
      () =>
        !disableARIA && {
          'aria-controls': accordionBodyId.current,
          'aria-expanded': open,
        },
      []
    );

    return (
      <div className={accordionClass}>
        <AccordionHeader
          disableIcon={disableIcon}
          focusable={focusable}
          alignIconRight={alignIconRight}
          disableCollapse={disableCollapse}
          accordionBodyId={accordionBodyId.current}
          accordionId={accordionID.current}
          iconType={iconType}
          iconColor={iconColor}
          title={title}
          customIcon={customIcon}
          isTitleBold={isTitleBold}
          open={open}
          onToggle={toggleAccordion}
          {...ARIA}
          aria-controls={accordionBodyId.current}
          aria-expanded={open}
          selected={selected}
          customContent={customContent}
          size={size}
        />
        <div
          className={accordionBodyClass}
          style={(autoSetBodyHeight ? styleWithHeight : style) as CSSProperties}
          ref={onInitRef}
          id={accordionBodyId.current}
          aria-labelledby={accordionID.current}
        >
          {children}
        </div>
      </div>
    );
  }
);

Accordion.displayName = 'Accordion';

export { Accordion };
