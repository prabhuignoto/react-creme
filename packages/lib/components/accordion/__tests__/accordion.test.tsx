import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { Accordion } from '../accordion';

describe('Accordion', () => {
  it('should render accordion', () => {
    const { container } = render(
      <Accordion>
        <p>this is a test</p>
      </Accordion>
    );

    expect(container.firstChild).toBeInTheDocument();
    expect(container.firstChild).toHaveClass('rc-accordion');
  });

  it('should render snapshot', () => {
    const { container } = render(
      <Accordion>
        <p>this is a test</p>
      </Accordion>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should toggle content', async () => {
    const { getByRole, container } = render(
      <Accordion>
        <p>this is a test</p>
      </Accordion>
    );

    await act(async () => {
      fireEvent.click(getByRole('button'));
    });

    await waitFor(async () => {
      expect(container.firstChild).toHaveClass('rc-accordion-open');
    });
  });

  it('should call onExpanded', () => {
    const onExpanded = jest.fn();

    const { getByRole } = render(
      <Accordion onExpanded={onExpanded}>
        <p>this is a test</p>
      </Accordion>
    );

    fireEvent.click(getByRole('button'));

    expect(onExpanded).toHaveBeenCalled();
  });
});
