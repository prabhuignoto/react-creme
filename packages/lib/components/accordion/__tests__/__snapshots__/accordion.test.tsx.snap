// Vitest Snapshot v1

exports[`Accordion > should render snapshot 1`] = `
<div
  class="rc-accordion rc-accordion-no-border"
  style="--max-height: 0px; --title-color: #000; --transition: cubic-bezier(0.19, 1, 0.22, 1);"
>
  <div
    aria-controls="accordion-body-uy3sRbP8GxciqfDFQc2Kt"
    aria-expanded="false"
    class="rc-accordion-header rc-accordion-focusable"
    id="accordion-MounNkzkgcJKtxMQ-Io9H"
    role="button"
    style="outline: none; position: relative;"
    tabindex="0"
  >
    <span
      class="rc-accordion-icon rc-accordion-icon-chevron"
    >
      <svg
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9 18l6-6-6-6"
        />
      </svg>
    </span>
    <span
      class="rc-accordion-title"
    />
    <span
      class="rc-focus-ring"
      style="width: 6px; height: 6px;"
    />
  </div>
  <div
    aria-labelledby="accordion-MounNkzkgcJKtxMQ-Io9H"
    class="rc-accordion-body rc-accordion-close"
    id="accordion-body-uy3sRbP8GxciqfDFQc2Kt"
    style="--max-height: 0px; --title-color: #000; --transition: cubic-bezier(0.19, 1, 0.22, 1);"
  >
    <p>
      this is a test
    </p>
  </div>
</div>
`;
